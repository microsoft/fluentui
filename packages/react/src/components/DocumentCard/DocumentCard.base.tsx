import * as React from 'react';
import { IProcessedStyleSet } from '../../Styling';
import {
  classNamesFunction,
  KeyCodes,
  getNativeProps,
  divProperties,
  warnDeprecations,
  initializeComponentRef,
} from '../../Utilities';
import {
  DocumentCardType,
  IDocumentCard,
  IDocumentCardProps,
  IDocumentCardStyleProps,
  IDocumentCardStyles,
} from './DocumentCard.types';

const getClassNames = classNamesFunction<IDocumentCardStyleProps, IDocumentCardStyles>();

const COMPONENT_NAME = 'DocumentCard';

/**
 * {@docCategory DocumentCard}
 */
export class DocumentCardBase extends React.Component<IDocumentCardProps, any> implements IDocumentCard {
  public static defaultProps: IDocumentCardProps = {
    type: DocumentCardType.normal,
  };

  private _rootElement = React.createRef<HTMLDivElement>();
  private _classNames: IProcessedStyleSet<IDocumentCardStyles>;

  constructor(props: IDocumentCardProps) {
    super(props);

    initializeComponentRef(this);
    warnDeprecations(COMPONENT_NAME, props, {
      accentColor: undefined,
    });
  }

  public render(): JSX.Element {
    // eslint-disable-next-line deprecation/deprecation
    const { onClick, onClickHref, children, type, accentColor, styles, theme, className } = this.props;
    const nativeProps = getNativeProps<React.HTMLAttributes<HTMLDivElement>>(this.props, divProperties, [
      'className',
      'onClick',
      'type',
      'role',
    ]);
    const actionable = onClick || onClickHref ? true : false;

    this._classNames = getClassNames(styles!, {
      theme: theme!,
      className,
      actionable,
      compact: type === DocumentCardType.compact ? true : false,
    });

    // Override the border color if an accent color was provided (compact card only)
    let style;
    if (type === DocumentCardType.compact && accentColor) {
      style = {
        borderBottomColor: accentColor,
      };
    }

    // if this element is actionable it should have an aria role
    const role = this.props.role || (actionable ? (onClick ? 'button' : 'link') : undefined);
    const tabIndex = actionable ? 0 : undefined;

    return (
      <div
        ref={this._rootElement}
        tabIndex={tabIndex}
        data-is-focusable={actionable}
        role={role}
        className={this._classNames.root}
        onKeyDown={actionable ? this._onKeyDown : undefined}
        onClick={actionable ? this._onClick : undefined}
        style={style}
        {...nativeProps}
      >
        {children}
      </div>
    );
  }

  public focus(): void {
    if (this._rootElement.current) {
      this._rootElement.current.focus();
    }
  }

  private _onClick = (ev: React.MouseEvent<HTMLElement>): void => {
    this._onAction(ev);
  };

  private _onKeyDown = (ev: React.KeyboardEvent<HTMLElement>): void => {
    // eslint-disable-next-line deprecation/deprecation
    if (ev.which === KeyCodes.enter || ev.which === KeyCodes.space) {
      this._onAction(ev);
    }
  };

  private _onAction = (ev: React.SyntheticEvent<HTMLElement>): void => {
    const { onClick, onClickHref, onClickTarget } = this.props;

    if (onClick) {
      onClick(ev);
    } else if (!onClick && onClickHref) {
      // If no onClick Function was provided and we do have an onClickHref, redirect to the onClickHref
      if (onClickTarget) {
        window.open(onClickHref, onClickTarget, 'noreferrer noopener nofollow');
      } else {
        window.location.href = onClickHref;
      }

      ev.preventDefault();
      ev.stopPropagation();
    }
  };
}
