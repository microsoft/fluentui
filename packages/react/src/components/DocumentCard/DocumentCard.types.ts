import * as React from 'react';
import type { IStyle, ITheme } from '../../Styling';
import type { IBaseProps, IRefObject, IStyleFunctionOrObject } from '../../Utilities';

/**
 * {@docCategory DocumentCard}
 */
export interface IDocumentCard {
  /**
   * Sets focus to the DocumentCard.
   */
  focus: () => void;
}

/**
 * {@docCategory DocumentCard}
 */
export interface IDocumentCardProps extends IBaseProps<IDocumentCard>, React.HTMLAttributes<HTMLDivElement> {
  /**
   * Optional callback to access the IDocumentCard interface. Use this instead of ref for accessing
   * the public methods and properties of the component.
   */
  componentRef?: IRefObject<IDocumentCard>;

  /**
   * The type of DocumentCard to display.
   * @defaultvalue DocumentCardType.normal
   */
  type?: DocumentCardType;

  /**
   * Function to call when the card is clicked or keyboard Enter/Space is pushed.
   */
  onClick?: (ev?: React.SyntheticEvent<HTMLElement>) => void;

  /**
   * A URL to navigate to when the card is clicked. If a function has also been provided,
   * it will be used instead of the URL.
   */
  onClickHref?: string;

  /**
   * A target browser context for opening the link. If not specified, will open in the same tab/window.
   */
  onClickTarget?: string;

  /**
   * Aria role assigned to the documentCard (Eg. button, link).
   * Use this to override the default assignment.
   * @defaultvalue When `onClick` is provided, default role will be 'button'.
   * When `onClickHref` is provided, default role will be 'link'.
   */
  role?: string;

  /**
   * Hex color value of the line below the card, which should correspond to the document type.
   * This should only be supplied when using the 'compact' card layout.
   *
   * @deprecated To be removed at \>= v5.0.0.
   */
  accentColor?: string;

  /**
   * Child components to render within the card.
   */
  children?: React.ReactNode;

  /**
   * Call to provide customized styling that will layer on top of the variant rules
   */
  styles?: IStyleFunctionOrObject<IDocumentCardStyleProps, IDocumentCardStyles>;

  /**
   * Theme provided by HOC.
   */
  theme?: ITheme;

  /**
   * Optional override class name
   */
  className?: string;
}

/**
 * {@docCategory DocumentCard}
 */
export enum DocumentCardType {
  /**
   * Standard DocumentCard.
   */
  normal = 0,
  /**
   * Compact layout. Displays the preview beside the details, rather than above.
   */
  compact = 1,
}

/**
 * {@docCategory DocumentCard}
 */
export interface IDocumentCardStyleProps {
  /**
   * Accept theme prop.
   */
  theme: ITheme;

  /**
   * Optional override class name
   */
  className?: string;

  /**
   * True when the card has a click action.
   */
  actionable?: boolean;

  /**
   * Compact variant of the card.
   */
  compact?: boolean;
}

/**
 * {@docCategory DocumentCard}
 */
export interface IDocumentCardStyles {
  root: IStyle;
}

/**
 * {@docCategory DocumentCard}
 */
export interface IDocumentCardContext {
  role?: string;
  tabIndex?: number;
}
