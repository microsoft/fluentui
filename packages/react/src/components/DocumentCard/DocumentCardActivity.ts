import * as React from 'react';
import { styled } from '../../Utilities';
import { DocumentCardActivityBase } from './DocumentCardActivity.base';
import { getStyles } from './DocumentCardActivity.styles';
import {
  IDocumentCardActivityProps,
  IDocumentCardActivityStyleProps,
  IDocumentCardActivityStyles,
} from './DocumentCardActivity.types';

export const DocumentCardActivity: React.FunctionComponent<IDocumentCardActivityProps> = styled<
  IDocumentCardActivityProps,
  IDocumentCardActivityStyleProps,
  IDocumentCardActivityStyles
>(DocumentCardActivityBase, getStyles, undefined, { scope: 'DocumentCardActivity' });
