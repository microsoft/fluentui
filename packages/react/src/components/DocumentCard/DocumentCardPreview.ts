import * as React from 'react';
import { styled } from '../../Utilities';
import { DocumentCardPreviewBase } from './DocumentCardPreview.base';
import { getStyles } from './DocumentCardPreview.styles';
import {
  IDocumentCardPreviewProps,
  IDocumentCardPreviewStyleProps,
  IDocumentCardPreviewStyles,
} from './DocumentCardPreview.types';

export const DocumentCardPreview: React.FunctionComponent<IDocumentCardPreviewProps> = styled<
  IDocumentCardPreviewProps,
  IDocumentCardPreviewStyleProps,
  IDocumentCardPreviewStyles
>(DocumentCardPreviewBase, getStyles, undefined, { scope: 'DocumentCardPreview' });
