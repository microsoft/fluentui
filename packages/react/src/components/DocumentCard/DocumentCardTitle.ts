import * as React from 'react';
import { styled } from '../../Utilities';
import { DocumentCardTitleBase } from './DocumentCardTitle.base';
import { getStyles } from './DocumentCardTitle.styles';
import {
  IDocumentCardTitleProps,
  IDocumentCardTitleStyleProps,
  IDocumentCardTitleStyles,
} from './DocumentCardTitle.types';

export const DocumentCardTitle: React.FunctionComponent<IDocumentCardTitleProps> = styled<
  IDocumentCardTitleProps,
  IDocumentCardTitleStyleProps,
  IDocumentCardTitleStyles
>(DocumentCardTitleBase, getStyles, undefined, { scope: 'DocumentCardTitle' });
