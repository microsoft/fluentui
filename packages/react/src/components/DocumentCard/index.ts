export * from './DocumentCard';
export * from './DocumentCard.types';
export * from './DocumentCardActions';
export * from './DocumentCardActions.types';
export * from './DocumentCardActivity';
export * from './DocumentCardActivity.types';
export * from './DocumentCardDetails';
export * from './DocumentCardDetails.types';
export * from './DocumentCardLocation';
export * from './DocumentCardLocation.types';
export * from './DocumentCardPreview';
export * from './DocumentCardPreview.types';
export * from './DocumentCardImage';
export * from './DocumentCardImage.types';
export * from './DocumentCardTitle';
export * from './DocumentCardTitle.types';
export * from './DocumentCardLogo';
export * from './DocumentCardLogo.types';
export * from './DocumentCardStatus';
export * from './DocumentCardStatus.types';
