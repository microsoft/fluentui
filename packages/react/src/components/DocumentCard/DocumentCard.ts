import * as React from 'react';
import { styled } from '../../Utilities';
import { DocumentCardBase } from './DocumentCard.base';
import { getStyles } from './DocumentCard.styles';
import { IDocumentCardProps, IDocumentCardStyleProps, IDocumentCardStyles } from './DocumentCard.types';

export const DocumentCard: React.FunctionComponent<IDocumentCardProps> = styled<
  IDocumentCardProps,
  IDocumentCardStyleProps,
  IDocumentCardStyles
>(DocumentCardBase, getStyles, undefined, { scope: 'DocumentCard' });
