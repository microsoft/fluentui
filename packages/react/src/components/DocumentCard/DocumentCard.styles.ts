import { getGlobalClassNames, getInputFocusStyle } from '../../Styling';
import { IsFocusVisibleClassName } from '../../Utilities';
import { IDocumentCardStyleProps, IDocumentCardStyles } from './DocumentCard.types';
import { DocumentCardPreviewGlobalClassNames as previewClassNames } from './DocumentCardPreview.styles';
import { DocumentCardActivityGlobalClassNames as activityClassNames } from './DocumentCardActivity.styles';
import { DocumentCardTitleGlobalClassNames as titleClassNames } from './DocumentCardTitle.styles';
import { DocumentCardLocationGlobalClassNames as locationClassNames } from './DocumentCardLocation.styles';

const GlobalClassNames = {
  root: 'ms-DocumentCard',
  rootActionable: 'ms-DocumentCard--actionable',
  rootCompact: 'ms-DocumentCard--compact',
};

export const getStyles = (props: IDocumentCardStyleProps): IDocumentCardStyles => {
  const { className, theme, actionable, compact } = props;
  const { palette, fonts, effects } = theme;

  const classNames = getGlobalClassNames(GlobalClassNames, theme);

  return {
    root: [
      classNames.root,
      {
        WebkitFontSmoothing: 'antialiased',
        backgroundColor: palette.white,
        border: `1px solid ${palette.neutralLight}`,
        maxWidth: '320px',
        minWidth: '206px',
        userSelect: 'none',
        position: 'relative',
        selectors: {
          ':focus': {
            outline: '0px solid',
          },
          [`.${IsFocusVisibleClassName} &:focus`]: getInputFocusStyle(palette.neutralSecondary, effects.roundedCorner2),
          [`.${locationClassNames.root} + .${titleClassNames.root}`]: {
            paddingTop: '4px',
          },
        },
      },
      actionable && [
        classNames.rootActionable,
        {
          selectors: {
            ':hover': {
              cursor: 'pointer',
              borderColor: palette.neutralTertiaryAlt,
            },
            ':hover:after': {
              content: '" "',
              position: 'absolute',
              top: 0,
              right: 0,
              bottom: 0,
              left: 0,
              border: `1px solid ${palette.neutralTertiaryAlt}`,
              pointerEvents: 'none',
            },
          },
        },
      ],
      compact && [
        classNames.rootCompact,
        {
          display: 'flex',
          maxWidth: '480px',
          height: '108px',
          selectors: {
            [`.${previewClassNames.root}`]: {
              borderRight: `1px solid ${palette.neutralLight}`,
              borderBottom: 0, // Remove the usual border from the preview
              maxHeight: '106px',
              maxWidth: '144px',
            },
            [`.${previewClassNames.icon}`]: {
              maxHeight: '32px',
              maxWidth: '32px',
            },
            [`.${activityClassNames.root}`]: {
              paddingBottom: '12px',
            },
            [`.${titleClassNames.root}`]: {
              paddingBottom: '12px 16px 8px 16px',
              fontSize: fonts.mediumPlus.fontSize,
              lineHeight: '16px',
            },
          },
        },
      ],
      className,
    ],
  };
};
