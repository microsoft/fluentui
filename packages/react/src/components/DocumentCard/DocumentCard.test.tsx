import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { DocumentCard } from './DocumentCard';
import { DocumentCardTitle } from './DocumentCardTitle';
import { DocumentCardPreview } from './DocumentCardPreview';
import { DocumentCardActivity } from './DocumentCardActivity';

describe('DocumentCard', () => {
  it('renders DocumentCard correctly', () => {
    const component = renderer.create(
      <DocumentCard>
        <DocumentCardPreview previewImages={[]} />
        <DocumentCardTitle title="" />
        <DocumentCardActivity activity="" people={[{ name: '', profileImageSrc: '' }]} />
      </DocumentCard>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
