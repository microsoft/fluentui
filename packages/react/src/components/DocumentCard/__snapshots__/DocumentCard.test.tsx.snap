// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentCard renders DocumentCard correctly 1`] = `
<div
  className=
      ms-DocumentCard
      {
        -webkit-font-smoothing: antialiased;
        background-color: #ffffff;
        border: 1px solid #edebe9;
        max-width: 320px;
        min-width: 206px;
        position: relative;
        user-select: none;
      }
      &:focus {
        outline: 0px solid;
      }
      .ms-Fabric--isFocusVisible &:focus {
        border-color: #605e5c;
      }
      .ms-Fabric--isFocusVisible &:focus:after {
        border-radius: 2px;
        border: 2px solid #605e5c;
        bottom: -1px;
        content: '';
        left: -1px;
        pointer-events: none;
        position: absolute;
        right: -1px;
        top: -1px;
      }
      @media screen and (-ms-high-contrast: active), (forced-colors: active){.ms-Fabric--isFocusVisible &:focus:after {
        border-color: Highlight;
      }
      & .ms-DocumentCardLocation + .ms-DocumentCardTitle {
        padding-top: 4px;
      }
  role="group"
>
  <div
    className=
        ms-DocumentCardPreview
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          background-color: #faf9f8;
          border-bottom: 1px solid #edebe9;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 12px;
          font-weight: 400;
          overflow: hidden;
          position: relative;
        }
  />
  <div
    className=
        ms-DocumentCardTitle
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #323130;
          display: block;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 18px;
          font-weight: 400;
          height: 38px;
          line-height: 21px;
          overflow: hidden;
          padding-bottom: 8px;
          padding-left: 16px;
          padding-right: 16px;
          padding-top: 8px;
          word-wrap: break-word;
        }
    title=""
  >
    
  </div>
  <div
    className=
        ms-DocumentCardActivity
        {
          padding-bottom: 8px;
          padding-left: 16px;
          padding-right: 16px;
          padding-top: 8px;
          position: relative;
        }
  >
    <div
      className=
          ms-DocumentCardActivity-avatars
          {
            height: 32px;
            margin-left: -2px;
          }
    >
      <div
        className=
            ms-DocumentCardActivity-avatar
            {
              display: inline-block;
              height: 32px;
              position: relative;
              text-align: center;
              vertical-align: top;
              width: 32px;
            }
            &:after {
              border-radius: 50%;
              border: 2px solid #ffffff;
              bottom: -1px;
              content: " ";
              left: -1px;
              position: absolute;
              right: -1px;
              top: -1px;
            }
      >
        <div
          className=
              ms-Persona-coin
              ms-Persona--size32
              {
                -moz-osx-font-smoothing: grayscale;
                -webkit-font-smoothing: antialiased;
                font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                font-size: 14px;
                font-weight: 400;
              }
          role="presentation"
        >
          <div
            className=
                ms-Persona-imageArea
                {
                  flex: 0 0 auto;
                  height: 32px;
                  position: relative;
                  text-align: center;
                  width: 32px;
                }
            role="presentation"
          >
            <div
              aria-hidden="true"
              className=
                  ms-Persona-initials
                  {
                    background-color: #0078D4;
                    border-radius: 50%;
                    color: #ffffff;
                    font-size: 14px;
                    font-weight: 600;
                    height: 32px;
                    line-height: 32px;
                  }
                  @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
                    -ms-high-contrast-adjust: none;
                    background-color: Window !important;
                    border: 1px solid WindowText;
                    box-sizing: border-box;
                    color: WindowText;
                    forced-color-adjust: none;
                  }
                  & i {
                    font-weight: 600;
                  }
            >
              <i
                aria-hidden={true}
                className=

                    {
                      -moz-osx-font-smoothing: grayscale;
                      -webkit-font-smoothing: antialiased;
                      display: inline-block;
                      font-family: "FabricMDL2Icons-0";
                      font-style: normal;
                      font-weight: normal;
                      speak: none;
                    }
                data-icon-name="Contact"
              >
                Óùª
              </i>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className=
          ms-DocumentCardActivity-details
          {
            height: 32px;
            left: 56px;
            position: absolute;
            top: 8px;
            width: calc(100% - 72px);
          }
    >
      <span
        className=
            ms-DocumentCardActivity-name
            {
              color: #323130;
              display: block;
              font-size: 12px;
              font-weight: 600;
              height: 15px;
              line-height: 15px;
              overflow: hidden;
              text-overflow: ellipsis;
              white-space: nowrap;
            }
      >
        
      </span>
      <span
        className=
            ms-DocumentCardActivity-activity
            {
              color: #605e5c;
              display: block;
              font-size: 12px;
              height: 15px;
              line-height: 15px;
              overflow: hidden;
              text-overflow: ellipsis;
              white-space: nowrap;
            }
      >
        
      </span>
    </div>
  </div>
</div>
`;
